Add validate(query:) method to Validatable protocol

odanylewycz opened the pull request:
Adds `validate(query:)` method to `Validatable` protocol for validating decoded query parameters (#2419). 

Previously, you could only validate the content that came from the request body. Now, this addition allows for validating decoded query parameters that conform to the Content protocol.

```swift
struct Hello: Content, Validatable {
    name: String?

    static func validations(_ validations: inout Validations) {
        validations.add("name", as: String.self, is: .alphanumeric)
    }
}
```

And the following URL, `http://localhost:8080/model?name=Vapor`, you can now validate the query parameter "name" as follows:

```swift
try Hello.validate(query: request)
let hello = try req.query.decode(Hello.self)
```

The existing content validation method `validate(_:)` has been renamed to `validate(content:)` to reduce ambiguity. 

```swift
try Hello.validate(content: request)
let hello = try req.content.decode(Hello.self)
```

0xTim commented:
@odanylewycz nice first-time PR! I've tidied up the comment a bit to use code formatting. Would you mind adding a test or two for the validation stuff?

odanylewycz commented:
@0xTim Thanks for the suggestion! I forgot to do it in the first place. I have gone ahead and added those tests as you requested. I reverted the prior commit because I misspelled "decoding", it was gonna bother me. Let me know if there is anything else I should add!

tanner0101 commented on Sources/Vapor/Validation/Validations.swift:
```diff
@@ -46,6 +46,14 @@ public struct Validations {
     }
     
+    public func validate(_ uri: URI) throws -> ValidationsResult {
+        
```
rm spaces (and one below)

odanylewycz replied:
Gotcha. I’ve usually added spaces above and below for readability, but I’ll go ahead and remove them. It’s a personal preference more than anything 😊

tanner0101 commented on Sources/Vapor/Validation/Validatable.swift:
```diff
@@ -17,6 +17,10 @@ extension Validatable {
         try self.validations().validate(request).assert()
     }
     
+    public static func validate(_ uri: URI) throws {
```
Since only the query string of the URI is being validated, maybe a label `query` would be useful here?

odanylewycz replied:
I only used “uri” as the label to match the type of the parameter. However, I believe a label of query would be more readable. So I’d agree with the change. Would you like me to modify the pull request?

MrLotU replied:
My personal suggestion would lean to `validateQuery(from uri: URI) throws {`.

As for how to proceed, you can push your updated version to your branch and it'll appear on GH so we can have another review round 😄

odanylewycz replied:
Thank you @MrLotU! The feedback is very helpful.

@tanner0101 After doing more review (being at my computer this time), I feel what best matches is the following syntax:

`validate(query: URI) throws {`

That is because I feel it follows the `validate(json: String) throws {` function in terms of style.

@MrLotU I feel `validateQuery(from uri: URI) throws {` is definitely more readable, but takes away from the current styling. I'd feel more inclined to follow that style ypu mentioned, if `validate(_ request: Request) throws {` was relabeled as `validateBody(from request: Request) throws {` as that method is really only validating the the body content, and not other parts of the request, such as the headers.

The way the decoding is set up, it looks like validating headers can be there in the future with some small changes, but as of now, the headers don't look like they are used during decoding, but rather only before to figure out the content type of the body coming in from the request.

So before I proceed, thoughts everyone?

tanner0101 replied:
Yeah I think `query` is better here. `from` makes sense in a vacuum, but since `validate(json:)` must have a clearer label (because it accepts "string" which is vague), then all having clear labels is more consistent.

tanner0101 replied:
FWIW, it might make sense to rename `validate(_ request: Request)` to `validate(content request: Request)`. You could also have `validate(query request: Request)` that calls `validate(query: request.url)` internally.

odanylewycz replied:
Agreed, for both points.

I'll edit `validate(_ request: Request)` to `validate(content request: Request)`. In addition, I did the same with `validate(query request: Request)`. I had to add a method of `validate(query: URI)` so I can add tests for it, otherwise I wasn't sure how to test it. If you have a better approach, please let me know, but this is what I've done so far.

`extension Validatable {
    public static func validate(content request: Request) throws {
        try self.validations().validate(request: request).assert()
    }
    
    public static func validate(query request: Request) throws {
        try self.validations().validate(query: request.url).assert()
    }
    
    public static func validate(json: String) throws {
        try self.validations().validate(json: json).assert()
    }
    
    public static func validate(query: URI) throws {
        try self.validations().validate(query: query).assert()
    } ...`

I fixed any references as well that resulted as build errors, in particular with `validate(content request: Request)`, the only one I was was in RouteTests.swift.

Once I get a thumbs up for changes, I'll then update the pull request.

tanner0101 replied:
LGTM :) Just be sure to add a deprecation for the removed `validate(_:Request)` method to https://github.com/vapor/vapor/tree/master/Sources/Vapor/Deprecations

odanylewycz replied:
Great! :) Thanks for telling me too! I would not have known to do that.

I have never also done this, so I am not sure if its right, would this be correct?

File name: Validatable+validate.swift
```
extension Validatable {
    @available(*, deprecated, message: "new required content parameter of type Request")
}
```
Thanks!

tanner0101 replied:
Try doing:
```swift
@available(*, deprecated, renamed: "validate(content:)")
```
That should give a nice message + an automatic fix it

odanylewycz replied:
Oh nice make sense. So just to be sure :)

```
extension Validatable {
    @available(*, deprecated, renamed: "validate(content:)")
    public static func validate(_ request: Request) -> throws {
        try self.validations().validate(request: request).assert()
    }
}
```

right? Sorry for so much validation (pun actually intended), first time I want to be sure its right! :)

odanylewycz replied:
Pull Request Updated.

tanner0101 replied:
You can just ignore that. Swift nightly is broken atm.

tanner0101 commented:
These changes are now available in [4.22.0](https://github.com/vapor/vapor/releases/tag/4.22.0)

tanner0101 commented:
@odanylewycz would you be open to submitting a PR to update https://github.com/vapor/docs/blob/master/4.0/docs/validation.md#validating-request with the new APIs added here?

odanylewycz commented:
@tanner0101 Certainly, it would be my pleasure! I'll get right on it.

odanylewycz commented:
Made the PR, let me know over on that PR if I need to change anything!
